libpod/boltdb_state.go

```diff
@@ -1101,6 +1107,382 @@ func (s *BoltState) PodContainers(pod *Pod) ([]*Container, error) {
 	return ctrs, nil
 }
 
+// AddVolume adds the given volume to the state. It allso adds ctrDepID to
+// the sub bucket holding the container dependencies that this volume has
+func (s *BoltState) AddVolume(volume *Volume, ctrDepID string) error {
+	if !s.valid {
+		return ErrDBBadConfig
+	}
+
+	if !volume.valid {
+		return ErrVolumeRemoved
+	}
+
+	volName := []byte(volume.Name())
+
+	volConfigJSON, err := json.Marshal(volume.config)
+	if err != nil {
+		return errors.Wrapf(err, "error marshalling volume %s config to JSON", volume.Name())
+	}
+
+	db, err := s.getDBCon()
+	if err != nil {
+		return err
+	}
+	defer s.closeDBCon(db)
+
+	err = db.Update(func(tx *bolt.Tx) error {
+		volBkt, err := getVolBucket(tx)
+		if err != nil {
+			return err
+		}
+
+		allVolsBkt, err := getAllVolsBucket(tx)
+		if err != nil {
+			return err
+		}
+
+		// Check if we already have a volume with the given name
+		volExists := allVolsBkt.Get(volName)
+		if volExists != nil {
+			return errors.Wrapf(ErrVolumeExists, "name %s is in use", volume.Name())
+		}
+
+		// We are good to add the volume
+		// Make a bucket for it
+		newVol, err := volBkt.CreateBucket(volName)
+		if err != nil {
+			return errors.Wrapf(err, "error creating bucket for volume %s", volume.Name())
+		}
+
+		// Make a subbucket for the containers using the volume
+		ctrDepsBkt, err := newVol.CreateBucket(volDependenciesBkt)
+		if err != nil {
+			return errors.Wrapf(err, "error creating bucket for containers using volume %s", volume.Name())
+		}
+		if ctrDepID != "" {
+			dependsCtr := []byte(ctrDepID)
+			if err := ctrDepsBkt.Put(dependsCtr, dependsCtr); err != nil {
+				return errors.Wrapf(err, "error storing container dependencies %q for volume %s in ctrDependencies bucket in DB", ctrDepID, volume.Name())
+			}
+		}
+
+		if err := newVol.Put(configKey, volConfigJSON); err != nil {
+			return errors.Wrapf(err, "error storing volume %s configuration in DB", volume.Name())
+		}
+
+		if err := allVolsBkt.Put(volName, volName); err != nil {
+			return errors.Wrapf(err, "error storing volume %s in all volumes bucket in DB", volume.Name())
+		}
+
+		return nil
+	})
+	return err
+}
+
+// UpdateVolume updates the container dependencies sub bucket of the given volume.
+// It either adds the ctrID if it is not found in the dependencies bucket or deletes
+// it from the bucket when found.
+// This is important as multiple containers can use one volume.
+func (s *BoltState) UpdateVolume(volume *Volume, ctrID string) error {
```

Comment:
Can we do this as a part of AddContainer and RemoveContainer instead? Check if the container depends on any volumes, and if so, grab the volume bucket and add/remove the dependency there?

Reply:
We could, but won't that be a bit misleading as the dependencies is a feature of the volume and not the container. Also I already go through the volumes when adding to figure out which ones are named volumes so I can set the source path, would probably have to do the same thing again on the database side to figure out which volume we should add the container ID as a dependency too.
If you really think it would be better that way, I can go ahead and change it.

Reply:
I think it's probably better to do it in the container commands - cleaner from an API standpoint, the DB will handle all the volume dependency work internally and libpod won't need to see it

Reply:
Okay, moved it to addContainer and removeContainer. I created RemoveVolCtrDep to remove the dependencies in the event of a force remove. Is it ok for the state function to have force as one of the function params, if so I can get rid of that function too.